### ISSUE TITLE
Ankidroid re-initializes WebView state for each review (so state is lost as cards flip)

### ISSUE THREAD

**Issue by armanshahrisa** [ankidroid/Anki-Android#4828]
###### Research
*Enter an [ x ] character to confirm the points below:*

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

###### More info
I have contacted Anki forum while ago asking them about multi-choice cards. A template was kindly provided by Damien Elmes. I have slightly modified it. Neither the original version nor the modified version is shown correctly in Ankidroid. Anki shows it perfectly by changing the answer to a bold green color after pressing "show answer". The Problem with Ankidroid is that the answers vanish upon pressing "show answer".

I have contacted the Ankidroid forum and Timothy Rae has pointed to code part where Ankidroid generates Answer:

https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fankidroid%2FAnki-Android%2Fblob%2Fmaster%2FAnkiDroid%2Fsrc%2Fmain%2Fjava%2Fcom%2Fichi2%2Fanki%2FAbstractFlashcardViewer.java%23L2065&sa=D&sntz=1&usg=AFQjCNGy-UYH97llPtSp3WPo-n4dUXxClg

The topic is here:
https://groups.google.com/forum/#!searchin/anki-android/sh88.arman%7Csort:date/anki-android/BGXfuhixpRw/vWNurTYFBQAJ

Then, I showed it to Damien Elmes through Anki support forum. He suggested that "it appears AnkiDroid is reloading the web view when revealing the answer, which means there is no way to carry state from the question side over to the answer side."

The topic is here:
https://anki.tenderapp.com/discussions/ankidesktop/27780-issue-with-multi-choice-cards

I convey the answer to Ankidroid discussion forum where I was suggested to create a topic here.

I'm very grateful for all your efforts in creating and maintaining Ankidroid. I would be even more thankful if somebody can help me overcome this problem.

Best regards,
Arman

###### Reproduction Steps

1. Create a template with the following information. 

```
{{Question}}



<center><div id=questions style="text-align:left;"></div></center>

<script>

function shuffle(array) {

var currentIndex = array.length, temporaryValue, randomIndex ;



// While there remain elements to shuffle...

while (0 !== currentIndex) {



// Pick a remaining element...

randomIndex = Math.floor(Math.random() * currentIndex);

currentIndex -= 1;



// And swap it with the current element.

temporaryValue = array[currentIndex];

array[currentIndex] = array[randomIndex];

array[randomIndex] = temporaryValue;

}



return array;

}



var right = "{{CorrectAnswer}}";

var all = [right, "{{WrongAnswer1}}", "{{WrongAnswer2}}", "{{WrongAnswer3}}"];

var buf = "<ol type=a>";

shuffle(all);

for (var i=0; i<4; i++) {

buf += "<li>"+all[i];

}

buf += "<ol type=a>";



document.getElementById("questions").innerHTML = buf;

</script>






Styling:

.card {

font-family: arial;

font-size: 20px;

text-align: left;

color: black;

background-color: white;

margin-left:auto; margin-right:auto; width:80%

}

img { margin-bottom: 2em; }



.field_set_col{

margin-bottom: 1em;

border: 2px solid #007868;

border-right: 2px solid #007868;

border-bottom: 2px solid #007868;

}





li{

margin: 20px 0;

}




Back template:

{{Question}}







<div id=ansArea></div>



<script>

var buf = "<ol type=a>";

var ans;

for (var i=0; i<4; i++) {

ans = all[i];

if (ans == right) {

ans = "<font color='#007868'><b>"+ans+"</font></b>";

}

buf += "<li>"+ans;

}

buf += "<ol type=a>";

document.getElementById("ansArea").innerHTML = buf;

</script>



<hr id=answer>



{{Extra}}

```

2. Create a multi-choice question and check it.

###### Expected Result
Answer must be shown in bold green color.

###### Actual Result
"show answer" will cause Anki to show the correct answer in bold green color but in Ankidroid, the answers are vanished upon pressing "show answer". 

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

**Comment by mikehardy** [comment]
This may be interesting - it's a polished project that does the web rendering work required for multiple choice, within the Anki sync ecosystem: https://github.com/qwiglydee/anki-interactive

**Comment by mikehardy** [comment]
Refining the name on this issue slightly after investigation in the area yesterday. We do use the same WebView so this is potentially fixable, but we use an API in the WebView for the card flip that definitely reinitializes state. Maybe re-writing the DOM or similar would allow state to be preserved. Not a high priority but it does mean this is more feasible than originally thought.

**Comment by mikehardy** [comment]
A bit more investigation on this related to #5213 - a possible re-implementation of how the reviewer loads content could be done by taking the 3 main loaded items (user question, user answer, user CSS), opening message channels between the Java app and the internal WebView (https://developer.android.com/reference/android/webkit/WebView.html#createWebMessageChannel()) and then using the messages to set .innerHTML on those three items.

This may have performance benefits as well, especially if one or more of the items didn't change or was complicated and had cacheable parts.

**Comment by sudomain** [comment]
This isn't something I've personally tested, but I just saw a reddit thread pertaining to a multiple choice template. [See here](https://gist.github.com/hgiesel/2e8361afccca5713414a6a4ee66b7ece) for information on the template. It uses [anki-persistence](https://github.com/SimonLammer/anki-persistence) to persist data between the front and back of a card. The README.md of this project appearsto show it working with AnkiDroid

**Comment by david-allison-1** [comment]
This is something I'd like to see fixed as there are decks on AnkiWeb which aren't compatible with AnkiDroid right now due to this (example: https://ankiweb.net/shared/info/920300037)

**Comment by mikehardy** [comment]
would tie with javascript hooks etc. I'm unsure how to do deal with this at all, as we do an HTML flip vs a card rewrite. I suppose you could just keep a good handle on the divs for the card and use the HTML DOM to rewrite them vs doing the whole card load? Heck, might even be faster.

**Comment by david-allison-1** [comment]
I think the answer is probably "see how Anki Desktop does it". It'll almost certainly be faster

**Comment by Rusca8** [comment]
Hi, just came to say that now Anki Desktop (2.1.45) changed the handling of Javascript even further apart from AnkiDroid.
Javascript variables and functions are kept on Desktop (it even raises an error if you declare the same varible with `let` both on `front` and `back` of the card). AnkiDroid seems to be generating everything from scratch when switching to `back`.

Depending on what you try to do with JS in your cards, this makes for a quite hard to overcome compatibility issue.

**Comment by mikehardy** [comment]
Oh, that is only unfortunate in that we are not behaving the same way. If Desktop is behaving the way you describe, they are doing it the correct way, I think.

It would be interesting to see exactly how they are doing this, are they re-writing the DOM somehow? However they are doing it, the closer we can get to doing the same in AnkiDroid, the better I think

**Comment by dae** [comment]
Anki's been reusing the same page since at least 2011. It used to use jQuery to do $(topelement).html(), which seemed to tolerate variable redefinition (maybe it ran the scripts in their own environment?), but I seem to recall even back then variables would persist, and certainly if they were set on `window`. As of a few versions ago, we now set innerHTML directly and run the scripts without the help of jQuery.

**Comment by mikehardy** [comment]
Cool - thanks for the detail @dae - always appreciated. As I do more web-tech dev work I slowly understand more and more of this stuff. Eventually I'll be able to perhaps actually implement the useful change here! Or if anyone else wants to take it up, it would be truly loved by all the people that we push towards Javascript add-ons in the future (or the early adopters playing with it now)

**Comment by yellowjello** [comment]
I might be able to take this on if I have a bit more time in the future.

But it looks like Anki Desktop updates the HTML in-place by calling a helper JS function via an eval. The WebView class has an `evaluateJavascript` method, so AnkiDroid can likely do the same as well.

We might even potentially be able to reuse much of Anki Desktop's JS helper code.

For those interested in working on this issue, I've created a simple deck for testing: [JSTest.apkg.zip](https://github.com/ankidroid/Anki-Android/files/10952338/JSTest.apkg.zip)

The cards show two randomly generated numbers on the front: A and B. After pressing "show answer", the expected behavior is for the value of A to change because the JS on the card should be rerun (which matches the desktop behavior) and the value of B should stay the same (to demonstrate state persistence)

<img width="358" alt="Screenshot 2023-03-12 at 4 04 03 PM" src="https://user-images.githubusercontent.com/5134058/224579658-f8745897-9694-4d3e-8923-334e9d26da36.png">

**Comment by BrayanDSO** [comment]
@dae, you are the person that knows better the Anki/Backend code, so I'm gonna ask for your advice about if it is possible to use the new backend for the reviewer webview logic/render

I've got some plans about improving the reviewer, so it would be nice to know if this is an option

**Comment by dae** [comment]
The plan is to make the reviewer usable on all the clients and move it into the shared ts/ area, but I haven't had a chance to do that yet. The existing code is mostly platform agnostic, and could be adapted fairly easy with copy+paste to use in AnkiDroid for now. When making it into a shared component in the future, we may want to revisit the public API, and consider switching state control from the host application to JS (eg, JS tells host to show answer buttons, instead of host telling JS to show answer). Or we may even want to move the button rendering into the same webview, so even more code can be shared.

**Comment by BrayanDSO** [comment]
> The existing code is mostly platform agnostic, and could be adapted fairly easy with copy+paste to use in AnkiDroid for now.

Lovely. It should already do wonders. I'll take a look at it after #14171 is merged.

> When making it into a shared component in the future, we may want to revisit the public API, and consider switching state control from the host application to JS (eg, JS tells host to show answer buttons, instead of host telling JS to show answer). Or we may even want to move the button rendering into the same webview, so even more code can be shared.

Sounds cool. It would help creating cross-platform JS addons in the future as well

**Comment by dae** [comment]
A couple of complicating factors to consider:

- AnkiDroid has no JS tooling at the moment, so you'd need to either (a) use the desktop's pre-made JS bundle, or (b) add that tooling to AnkiDroid.  The former would require changes to the backend release so the file gets included, the latter would require adding node tooling to either Droid or Backend.
- The type-in-the-answer handling is likely different, so a bit of work would be required to adapt it. AnkiDroid appears to have two separate ways to type in the answer at the moment which further complicates things; I'd suggest dropping one of the options in the future.

**Comment by mikehardy** [comment]
We're already going to have some node tooling for our crowdin v2 API sync process - it is currently slated just for the github workflows area but it *will* exist, no reason we can't have a shared blob of it and re-use it for app "render" (conversion of TS/JS/whatever to a loadable JS bundle)

**Comment by BrayanDSO** [comment]
Closing this in favor of #14302
